### PR TITLE
Migrate Windows to fuse3 API

### PR DESCRIPTION
**Pull request by infeo**
This PR migrates the windows jfuse project to use fuse3 instead of fuse2.

It had to be adjusted to winfsp specific quirks:
* Winfsp uses [fuse-3.2](https://github.com/winfsp/winfsp/blob/deee7eddedb24ba58c741cbb1421baf2e37fe5d4/inc/fuse3/fuse_common.h#L39-L40)
* even though the fuse header specifies the `fuse_loop_mt()` function, jextract was unable to extract it and `fuse_loop_mt_31()` had to be used
* to correctly call the fuse3 functions, the explicit fuse names in winfsp (`fuse3_...`) had to be used and redefinition is skipped by defining `WINFSP_DLL_INTERNAL`
* in fuse3 `fuse_parse_cmd_line(...)` is not defined. Looking at the source code it is used nonetheless:
https://github.com/winfsp/winfsp/blob/8d83d46e7f12fc1cb343f7917c6c59ee3a8fd0b6/src/dll/fuse3/fuse3.c#L37
  This works due to a longer chain of imports and since Winfsp is using this construct, i decided to also use it by adding an additonal jextract in an extra package

### CODE REVIEW

File: jfuse-win-amd64/pom.xml [hunk 1]
```diff
@@ -88,27 +88,55 @@
...
-										<includeStruct>fuse_operations</includeStruct>
-										<includeStruct>fuse_file_info</includeStruct>
+										<includeStruct>fuse3_operations</includeStruct>
+										<includeStruct>fuse3_file_info</includeStruct>
 										<includeStruct>fuse_stat</includeStruct>
 										<includeStruct>fuse_statvfs</includeStruct>
 										<includeStruct>fuse_timespec</includeStruct>
-										<includeStruct>fuse_conn_info</includeStruct>
+										<includeStruct>fuse3_conn_info</includeStruct>
+										<includeStruct>fuse3_loop_config</includeStruct>
```
Comment (overheadhunter) [inline jfuse-win-amd64/pom.xml:118]: `fuse3_loop_config` is unused, isn't it?

File: jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/extr/fuse3_loop_config.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+// Generated by jextract
+
+package org.cryptomator.jfuse.win.amd64.extr;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
+import java.lang.foreign.*;
+import static java.lang.foreign.ValueLayout.*;
+public class fuse3_loop_config {
```
Comment (overheadhunter) [inline jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/extr/fuse3_loop_config.java:10]: unused?

**Review by overheadhunter — CHANGES_REQUESTED** [review]
Good work, just found a few opportunities for cleanup.

File: jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseMountImpl.java [hunk 3]
```diff
@@ -5,23 +5,26 @@
...
 
-record FuseMountImpl(MemoryAddress fuse, MemoryAddress ch, FuseArgs args) implements FuseMount {
+record FuseMountImpl(MemoryAddress fuse, FuseArgs fuseArgs) implements FuseMount {
 
 	@Override
 	public int loop() {
-		// TODO support fuse_loop_mt if args.multiThreaded()
-		return fuse_h.fuse_loop(fuse);
+		if (fuseArgs.multiThreaded()) {
+			return fuse_h.fuse3_loop_mt_31(fuse, 0);
```
[the remark this answers is not in the document]
Comment (infeo) [reply jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseMountImpl.java:13]: It is hard to document, if you just followed the linux implementaiton 😅 

The parameter is called `clone_fd`. In fuse3.2 it changed to a struct containg a field representing it , see http://libfuse.github.io/doxygen/structfuse__loop__config.html#a6a8527939446625e5466cd8baf399f08
```c
struct fuse_loop_config {
         int clone_fd;
  
         unsigned int max_idle_threads;
};
```

[the remark this answers is not in the document]
Comment (overheadhunter) [reply jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseMountImpl.java:13]: Oh I just realized, that the jextracted function uses the name `clone_fd` as well. I thought it was just some anonymous int. In that case, ignore my comment. We should be able to google the meaning of `clone_fd` any time, no need to document it at this place.

File: jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/extr/fuse3_loop_config.java (hunk 2 above)
Comment (infeo) [reply jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/extr/fuse3_loop_config.java]: Fixed in https://github.com/cryptomator/jfuse/pull/7/commits/2ae3e52f33e581397708336f986743dfca589850

File: jfuse-win-amd64/pom.xml (hunk 1 above)
Comment (infeo) [reply jfuse-win-amd64/pom.xml]: Fixed in https://github.com/cryptomator/jfuse/pull/7/commits/2ae3e52f33e581397708336f986743dfca589850

File: jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseImpl.java [hunk 4]
```diff
@@ -128,8 +121,10 @@ private int access(MemoryAddress path, int mask) {
 		return delegate.access(path.getUtf8String(0), mask);
```
[the remark this answers is not in the document]
Comment (infeo) [reply jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseImpl.java]: Implemented in https://github.com/cryptomator/jfuse/pull/7/commits/2ae3e52f33e581397708336f986743dfca589850

File: jfuse-win-amd64/pom.xml [hunk 5]
```diff
@@ -88,27 +88,55 @@
...
+							<!-- winfsp does not support fuse_lowlevel.h but uses fuse2.h -->
+							<execution>
+								<id>jextract-fuse-parse-cmd-line</id>
+								<goals>
+									<goal>sources</goal>
+								</goals>
+								<configuration>
+									<targetPackage>org.cryptomator.jfuse.win.amd64.extr_fuse2</targetPackage>
+									<headerFile>${project.parent.basedir}/winfsp/inc/fuse/fuse_common.h</headerFile>
+									<headerClassName>fuse_2_h</headerClassName>
```
[the remark this answers is not in the document]
Comment (infeo) [reply jfuse-win-amd64/pom.xml]: Changed in https://github.com/cryptomator/jfuse/pull/7/commits/2ae3e52f33e581397708336f986743dfca589850

File: jfuse-win-amd64/pom.xml [hunk 6]
```diff
@@ -88,27 +88,55 @@
...
+								</configuration>
+							</execution>
+							<!-- winfsp does not support fuse_lowlevel.h but uses fuse2.h -->
+							<execution>
+								<id>jextract-fuse-parse-cmd-line</id>
+								<goals>
+									<goal>sources</goal>
+								</goals>
+								<configuration>
+									<targetPackage>org.cryptomator.jfuse.win.amd64.extr_fuse2</targetPackage>
```
[the remark this answers is not in the document]
Comment (infeo) [reply jfuse-win-amd64/pom.xml]: Changed in https://github.com/cryptomator/jfuse/pull/7/commits/2ae3e52f33e581397708336f986743dfca589850

File: jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseImpl.java [hunk 7]
```diff
@@ -239,21 +236,21 @@ private int unlink(MemoryAddress path) {
...
 	@VisibleForTesting
-	int utimens(MemoryAddress path, MemoryAddress times) {
+	int utimens(MemoryAddress path, MemoryAddress times, MemoryAddress fi) {
 		try (var scope = MemorySession.openConfined()) {
 			if (MemoryAddress.NULL.equals(times)) {
 				// set both times to current time (using on-heap memory segments)
 				var segment = MemorySegment.allocateNative(fuse_timespec.$LAYOUT().byteSize(), scope);
 				fuse_timespec.tv_sec$set(segment, 0);
-				fuse_timespec.tv_nsec$set(segment, 0); // FIXME use hard-coded UTIME_NOW
+				fuse_timespec.tv_nsec$set(segment, 0); //TODO: use something like stat_h.UTIME_NOW())
```
[the remark this answers is not in the document]
Comment (infeo) [reply jfuse-win-amd64/src/main/java/org/cryptomator/jfuse/win/amd64/FuseImpl.java]: i reverted the comment to its original in https://github.com/cryptomator/jfuse/pull/7/commits/a0b83c8c2745c89fb3cea2f65e36f169b652534d

**Review by overheadhunter — APPROVED** [review]

**Pull request merged**